### PR TITLE
Defer context menus on gtk and windows.

### PR DESCRIPTION
**Pull request by jneem**
Windows is currently untested -- I'll check it if CI passes.

### CODE REVIEW

**Comment by cmyr** [comment]
Tested on windows; I don't get the crash I would get otherwise, but i do get lots of warnings like,

```
WARN  [druid_shell::platform::windows::window] failed to release mouse capture: HRESULT 0x0: The operation completed successfully.
```

Not sure if this is a concern?

**Comment by jneem** [comment]
I'm not sure either, but I noticed that also. I haven't had time to dig in yet, though.

**Comment by jneem** [comment]
Ok, so I've disabled the warning. I'm not completely confident that it's the right thing, but my understanding of the situation is:
- the mouse goes down, and so we capture it
- the mouse goes up, and so we schedule a menu to appear when the handler is dropped. We also see that all mouse buttons were released, so we schedule a call to release the capture.
- we drop the handle, and run the context menu. While the menu is being run, the capture gets released somehow.
- we finish running the menu, and try to drop the capture but we get an error because there isn't a mouse capture active.

I'm not sure if there is much to be done here besides disabling the warning when we try to release a non-existent capture. According to the previous comments in the source, we need to defer the capture. And once we've done that, there isn't really any way to stop the capture being released before we actually run the deferred thing.

**Comment by cmyr** [comment]
I guess I don't even really understand what the capture is doing, in this case? Maybe whoever is managing the menu handles capture?

**Comment by jneem** [comment]
Yeah, it's pretty mysterious to me. The usual code path is that we take capture when the first button goes down and release it when the last button goes up. But when the menu pops up in between those two things happening, I think that we lose control over the event loop? And then I wasn't able to track what's going on. Maybe someone more familiar with windows programming can weigh in...

File: druid-shell/src/platform/windows/window.rs [hunk 1]
```diff
@@ -1003,25 +1060,13 @@ impl WndProc for MyWndProc {
                             s.handler.mouse_down(&event);
                         } else {
                             s.handler.mouse_up(&event);
-                            should_release_capture = s.exit_mouse_capture(button);
+                            if s.exit_mouse_capture(button) {
+                                self.handle.borrow().defer(DeferredOp::ReleaseMouseCapture);
+                            }
```
Comment (cmyr) [inline druid-shell/src/platform/windows/window.rs:1065]: I'm curious what happens if we continue to release here like normal? Would we have already handled the deferred `ContextMenu` action?

Comment (jneem) [reply druid-shell/src/platform/windows/window.rs:1065]: Prior to this PR, we would do the context menu synchronously in `handler.mouse_up` above. As it is in this PR right now, we run the context menu just below this, at the closing brace on line 1067.

The latest changes to `ReleaseCapture` don't actually change the ordering between the context menu and releasing the capture. Before the changes, we ran the context menu when dropping the handle borrow, and then a few lines later we have the release capture. After the changes, we run the context menu *and* the release capture (in that order) when dropping the handle borrow. The main reason for the changes to `ReleaseCapture` in the latest patch is that we have a general mechanism for deferring now, so I figured we should use it. The only actual behavior change is the silencing of the warning.

Comment (cmyr) [reply druid-shell/src/platform/windows/window.rs:1065]: i'll experiment a bit.

**Review by cmyr — APPROVED** [review]
Okay after closer reading I think we were just incorrect to treat the `0` return value as an error. I'm happy with this going in, and we will continue to iterate in any case. Thanks!

**Pull request merged**